flink-runtime/src/test/java/org/apache/flink/runtime/clusterframework/TaskExecutorResourceUtilsTest.java

```diff
@@ -0,0 +1,323 @@
+		MemorySize.parse("1m"),
+		MemorySize.parse("2m"),
+		MemorySize.parse("3m"),
+		MemorySize.parse("4m"),
+		MemorySize.parse("5m"),
+		MemorySize.parse("6m"),
+		MemorySize.parse("7m"),
+		MemorySize.parse("8m"));
+
+	@Test
+	public void testGenerateDynamicConfigurations() {
+		String dynamicConfigsStr = TaskExecutorResourceUtils.generateDynamicConfigsStr(TM_RESOURCE_SPEC);
+		Map<String, String> configs = new HashMap<>();
+		for (String configStr : dynamicConfigsStr.split(" ")) {
+			assertThat(configStr, startsWith("-D"));
+			String[] configKV = configStr.substring(2).split("=");
+			assertThat(configKV.length, is(2));
+			configs.put(configKV[0], configKV[1]);
+		}
+
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.FRAMEWORK_HEAP_MEMORY.key())), is(TM_RESOURCE_SPEC.getFrameworkHeapSize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.TASK_HEAP_MEMORY.key())), is(TM_RESOURCE_SPEC.getTaskHeapSize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.TASK_OFF_HEAP_MEMORY.key())), is(TM_RESOURCE_SPEC.getTaskOffHeapSize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.SHUFFLE_MEMORY_MAX.key())), is(TM_RESOURCE_SPEC.getShuffleMemSize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.SHUFFLE_MEMORY_MIN.key())), is(TM_RESOURCE_SPEC.getShuffleMemSize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.MANAGED_MEMORY_SIZE.key())), is(TM_RESOURCE_SPEC.getManagedMemorySize()));
+		assertThat(MemorySize.parse(configs.get(TaskManagerOptions.MANAGED_MEMORY_OFFHEAP_SIZE.key())), is(TM_RESOURCE_SPEC.getOffHeapManagedMemorySize()));
+	}
+
+	@Test
+	public void testGenerateJvmParameters() throws Exception {
+		String jvmParamsStr = TaskExecutorResourceUtils.generateJvmParametersStr(TM_RESOURCE_SPEC);
+		MemorySize heapSizeMax = null;
+		MemorySize heapSizeMin = null;
+		MemorySize directSize = null;
+		MemorySize metaspaceSize = null;
+		for (String paramStr : jvmParamsStr.split(" ")) {
+			if (paramStr.startsWith("-Xmx")) {
+				heapSizeMax = MemorySize.parse(paramStr.substring("-Xmx".length()));
+			} else if (paramStr.startsWith("-Xms")) {
+				heapSizeMin = MemorySize.parse(paramStr.substring("-Xms".length()));
+			} else if (paramStr.startsWith("-XX:MaxDirectMemorySize=")) {
+				directSize = MemorySize.parse(paramStr.substring("-XX:MaxDirectMemorySize=".length()));
+			} else if (paramStr.startsWith("-XX:MetaspaceSize=")) {
+				metaspaceSize = MemorySize.parse(paramStr.substring("-XX:MetaspaceSize=".length()));
+			} else {
+				throw new Exception("Unknown JVM parameter: " + paramStr);
+			}
+		}
+
+		assertThat(heapSizeMax, is(TM_RESOURCE_SPEC.getFrameworkHeapSize().add(TM_RESOURCE_SPEC.getTaskHeapSize()).add(TM_RESOURCE_SPEC.getOnHeapManagedMemorySize())));
+		assertThat(heapSizeMin, is(heapSizeMax));
+		assertThat(directSize, is(TM_RESOURCE_SPEC.getTaskOffHeapSize().add(TM_RESOURCE_SPEC.getShuffleMemSize())));
```

Comment:
off heap managed memory is missing

Reply:
That uses unsafe memory, which should not be accounted for either `-Xmx` nor `-XX:MaxDirectMemorySize`.

Reply:
Yes you are right. Thanks for the clarification.